README.md
# Human Interval
Human-readable interval parser for Javascript.

Converts words written in English to numbers by using [node-numbered](https://github.com/blakeembrey/node-numbered).

Originally inspired by [matthewmueller/date](http://github.com/matthewmueller/date).

## Uses

Human Interval is used by job scheduling libraries such as [Agenda](https://github.com/Agenda/agenda#readme) and [Bree](https://jobscheduler.net). They are a job schedulers for Node.js with cron expression syntax, human-friendly times, Dates, and more!

## Example usage

```js
const humanInterval = require('human-interval');

setTimeout(() => {
  // Do something!
}, humanInterval('three minutes'));
```

## More sophisticated examples

Human Interval understands all of the following examples:

```js
humanInterval('minute');
humanInterval('one minute');
humanInterval('1.5 minutes');
humanInterval('3 days and 4 hours');
humanInterval('3 days, 4 hours and 36 seconds');
humanInterval('4 months, 3 days, 5 hours and forty-five seconds');
```

## The full list

### Units

Supports the following units in the plural and singular forms:

- `seconds`
- `minutes`
- `hours`
- `days`
- `weeks`
- `months` — assumes 30 days
- `years` — assumes 365 days

### Wordy numbers

Supports numbers being written out in English words.

```js
humanInterval('five minutes');
```

### Hyphenated numbers

Supports hyphenated numbers.

```js
humanInterval('twenty-five seconds');
```

### Negative numbers

Supports negative numbers if the time starts with a `-` symbol immediately followed by a number.

```js
humanInterval('-2 minutes');
```

# License
[The MIT License](LICENSE.md)


History.md
2.0.1 / 2021-01-16
==================

* Export types ([#43](https://github.com/agenda/human-interval/pull/43))
* Updated dev dependencies.

2.0.0 / 2020-10-16
==================

* [Refactor](https://github.com/agenda/human-interval/pull/37) human interval using [numbered](https://www.npmjs.com/package/numbered) package. Thank you @sampathBlam and @alFReD-NSH for your contributions!
  - Supports all the formats as previously, and more!
  - Supports numbers written as English words (one, two hundred)
  - Supports time expressions in singular and plural (minute and minutes)
  - Supports negative numbers (-2)
  - Supports hyphenated words (twenty-five)
* Updated dev dependencies.
* Added more tests.

**Breaking**

* Stopped testing Node.js 8 — might still work of course. We're now testing for Node.js v10, v12, and v14.

1.0.0 / 2019-11-24
==================

**Breaking**

* Remove Component.js support — this tool has not been maintained since 2015 https://github.com/agenda/human-interval/pull/10
* Drop Node.js 4, 5, and 6 support — might still work but we're not testing for these anymore

**Internal development tooling**

* Update development dependencies
* Add MIT license
* Switch testing framework to Ava from Mocha https://github.com/agenda/human-interval/pull/16
* Use XO for linting

0.1.6 / 2016-05-11
==================

* Fix bug for invalid input strings.

0.1.5 / 2015-07-30
==================

* Fixed a bug with commas, all tests are now passing.

0.1.4 / 2014-04-06
==================

* Add support for number omission.

0.1.3 / 2014-01-04
==================

* Added support for months.

0.1.2 / 2014-01-03
==================

* Added support for 'week' unit.

0.1.1 / 2013-10-28
==================

* Added better undefined handling.

0.1.0 / 2013-10-25
==================

* Initial Release.


test.js
const test = require('ava');
const humanInterval = require('.');

const macro = (t, inputs, expected) => {
  inputs = Array.isArray(inputs) ? inputs : [inputs];
  inputs.forEach(input => t.is(humanInterval(input), expected));
};

const units = {};
units.second = 1000;
units.minute = units.second * 60;
units.hour = units.minute * 60;
units.day = units.hour * 24;
units.week = units.day * 7;
units.month = units.day * 30;
units.year = units.day * 365;

const timeUnits = [
  'second',
  'minute',
  'hour',
  'day',
  'week',
  'month',
  'year'
];

const englishNumbers = [
  'one',
  'two',
  'three',
  'four',
  'five',
  'six',
  'seven',
  'eight',
  'nine',
  'ten'
];

// Invalid values:
test('Returns undefined when given undefined', macro, undefined, undefined);
test('Returns null when given null', macro, null, null);
test('Returns empty string when given empty string', macro, '', '');
test('Returns NaN when given unknown string', macro, 'foobar', Number.NaN);

// Single values
test('Returns the number when given a number', macro, 5000, 5000);
test('Understands numbers', macro, '2 seconds', 2 * units.second);
test('Understands decimals', macro, '2.5 seconds', 2.5 * units.second);
test('Understands long decimals', macro, '2.5555 seconds', 2.5555 * units.second);

// Understands time base units without numbers
// i.e. "second === 1000"
timeUnits.forEach(unit => {
  test(
    'Understands "' + unit + '" without number',
    macro,
    unit,
    Number(units[unit])
  );
});

// Understands time base units in singular
// i.e. "1 second === 1000"
timeUnits.forEach(unit => {
  test(
    'Understands "1 ' + unit + '"',
    macro,
    '1 ' + unit,
    Number(units[unit])
  );
});

// Understands time base units in plural
// i.e. "2 seconds === 2000"
timeUnits.forEach(unit => {
  test(
    'Understands ' + unit + 's in plural',
    macro,
    '2 ' + unit + 's',
    2 * Number(units[unit])
  );
});

// Understands time base units in decimals values
// i.e. "1.5 seconds === 1500"
timeUnits.forEach(unit => {
  test(
    'Understands ' + unit + 's in decimals',
    macro,
    '1.5 ' + unit + 's',
    1.5 * Number(units[unit])
  );
});

// English numbers
// i.e. "one second === 1000"
englishNumbers.forEach((number, i) => {
  test(
    'Understands english number ' + number,
    macro,
    number + ' seconds',
    (i + 1) * units.second
  );
});

// Understands negative numbers
// i.e. "-2 seconds === -2000"
timeUnits.forEach(unit => {
  test(
    'Understands negative ' + unit + 's',
    macro,
    '-2 ' + unit + 's',
    -2 * Number(units[unit])
  );
});

// Mixed
test('Understands mixed units', macro, '3 minutes and 30 seconds', (3 * units.minute) + (30 * units.second));
test('Understands mixed time expressions with plurals', macro, ['three minutes and 30 seconds', 'three minutes 30 seconds'], (3 * units.minute) + (30 * units.second));
test('Understands mixed time expressions with singulars', macro, ['one minute and 1 second', 'one minute 1 second'], units.minute + units.second);
test('Understands 2 digit english numbers', macro, 'thirty three seconds', 33 * units.second);
test('Understands mix units + multi digit english numbers', macro, 'hundred and three seconds and twelve minutes', (103 * units.second) + (12 * units.minute));
test('Understands hyphenated numbers', macro, 'three hundred and twenty-five seconds', (325 * units.second));


test/human-interval.js
var expect = require('expect.js'),
    humanInterval = require('../index.js');

describe('Human Interval', function() {
  it('returns the number when given a number', function() {
    expect(humanInterval(5000)).to.be(5000);
  });

  it('returns undefined when given undefined', function() {
    expect(humanInterval(undefined)).to.be(undefined);
  });

  it('does not require a number', function() {
    expect(humanInterval('week')).to.be(7 * 86400000);
  });

  describe('basic units', function() {
    it('understands seconds', function() {
      expect(humanInterval('1 second')).to.be(1000);
    });
    it('understands minutes', function() {
      expect(humanInterval('1 minute')).to.be(60000);
    });
    it('understands hours', function() {
      expect(humanInterval('1 hour')).to.be(3600000);
    });
    it('understands days', function() {
      expect(humanInterval('1 day')).to.be(86400000);
    });
    it('understands weeks', function() {
      expect(humanInterval('1 week')).to.be(7 * 86400000);
    });
    it('understands months', function() {
      expect(humanInterval('1 month')).to.be(30 * 86400000);
    });
    it('understands years', function() {
      expect(humanInterval('1 year')).to.be(31536000000);
    });
  });

  describe('basic numbers', function() {
    it('understands numbers', function() {
      expect(humanInterval('2 seconds')).to.be(2000);
    });

    it('understands decimals', function() {
      expect(humanInterval('2.5 seconds')).to.be(2500);
    });
  });

  describe('english numbers', function() {
    it('understands numbers', function() {
      expect(humanInterval('two seconds')).to.be(2000);
    });
  });

  describe('mixes', function() {
    it('works with long numbers', function() {
      expect(humanInterval('3 minutes and 30 seconds')).to.be(210000);
    });

    it('works with mixed units', function() {
      expect(humanInterval('3 minutes and 30 seconds')).to.be(210000);
    });

    it('works with mixed time expressions', function() {
      expect(humanInterval('three minutes and 30 seconds')).to.be(210000);
      expect(humanInterval('three minutes 30 seconds')).to.be(210000);
    });
  });

  describe('Error handling', function() {
    it('throw error on unrecognised time unit', function() {
      expect(humanInterval).withArgs('3 unrecognised').to.throwError();
      expect(humanInterval).withArgs('3 seconds').to.not.throwError();
    });
  });
});


index.js
const numbered = require('numbered');

const units = {};
units.second = 1000;
units.minute = units.second * 60;
units.hour = units.minute * 60;
units.day = units.hour * 24;
units.week = units.day * 7;
units.month = units.day * 30;
units.year = units.day * 365;

const regexp = /(second|minute|hour|day|week|month|year)s?/;

const humanInterval = time => {
  if (!time || typeof time === 'number') {
    return time;
  }

  let result = Number.NaN;

  time = time.replace(/([^a-z\d.-]|and)+/g, ' ');

  for (;;) {
    const match = time.match(regexp);
    if (!match) {
      return result;
    }

    const matchedNumber = time.slice(0, match.index).trim();
    const unit = units[match[1]];
    let number = 1;
    if (matchedNumber.length > 0) {
      number = Number.parseFloat(matchedNumber);
      if (Number.isNaN(number)) {
        number = numbered.parse(matchedNumber);
      }
    }

    if (Number.isNaN(result)) {
      result = 0;
    }

    result += number * unit;
    time = time.slice(match.index + match[0].length);
  }
};

module.exports = humanInterval;
